Update @wordpress/data to version 4.23.0.

techanvil opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #1769 

## Relevant technical choices

- The test fixes all relate to needing to wait for resolvers which are now asynchronous, since the introduction of the async resolver cache in `@wordpress/data` 4.23.0 - see here for more details: https://github.com/google/site-kit-wp/issues/1769#issuecomment-1221335331

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

techanvil commented on assets/js/components/surveys/SurveyViewTrigger.js:
```diff
@@ -37,11 +41,11 @@ export default function SurveyViewTrigger( { triggerID, ttl = 0 } ) {
 	const { triggerSurvey } = useDispatch( CORE_USER );
 
-	useMount( () => {
+	useEffect( () => {
```
See associated commit for details on this one - https://github.com/google/site-kit-wp/pull/6203/commits/cced86f665a7dba9fd56d46c9858f43e45050226

techanvil commented on assets/js/components/adminbar/AdminBarWidgets.test.js:
```diff
@@ -53,14 +53,14 @@ describe( 'AdminBarWidgets', () => {
 		);
 
-		await waitForRegistry();
+		await waitForResolvingPromise();
```
Some resolvers were still asynchronously resolving after `waitForRegistry`, causing an error to be thrown due to the component updating outside of an `act`. Waiting for active Promises to resolve, via awaiting a new Promise which resolves in a timeout, solves this.

aaemnnosttv replied:
In this situation I think it's still appropriate to wait for the registry as before. Ideally we could update `waitForRegistry` to wait for any remaining async updates but I realize that may not be possible.

techanvil replied:
I've taken a different approach to the waits, having added a wrapper to `setTimeout` that allows timeouts to be tracked and therefore waited for in a more deterministic manner.

Then created a new helper, `waitForRegistryAndTimeouts` which wraps both `waitForRegistry` and the new `waitForTimeouts` utility, and used this where `waitForResolvingPromise` was previously being used.

techanvil commented on assets/js/googlesitekit/datastore/user/feature-tours.js:
```diff
@@ -231,8 +231,8 @@ const baseControls = {
-				const initialVersion = registry
-					.select( CORE_USER )
+				const initialVersion = await registry
+					.__experimentalResolveSelect( CORE_USER )
```
As discussed [above](https://github.com/google/site-kit-wp/pull/6203/files#r1051087530), this is necessary to fix an issue with referring to resolver-derived values in a `useMount`. The `useMount` call that results in this control running is here:

https://github.com/google/site-kit-wp/blob/d0cd49c2d47d758ce7991fd7778a35bb755252a3/assets/js/components/FeatureTours.js#L37-L39

aaemnnosttv commented on tests/js/utils.js:
```diff
@@ -538,3 +538,14 @@ export const unexpectedSuccess = () => {
+ * @param {number} [delay] The delay in milliseconds. Defaults to 1.
+ * @return {Promise} A promise that resolves after the specified delay.
+ */
+export const resolvingPromise = ( delay = 1 ) =>
```
Is the delay particularly significant? I see in some tests we use `5` while others use the default. The current/closest use of this does not provide a delay at all https://github.com/google/site-kit-wp/blob/d0cd49c2d47d758ce7991fd7778a35bb755252a3/tests/js/setup-before-after.js#L63

Any reason not to use `undefined` as default here as well?

Secondarily, the name is a bit ambiguous. Maybe something like `waitForDelay` would be more appropriate? That would make the argument a bit more clear as well when used.

aaemnnosttv commented on tests/js/test-utils.js:
```diff
@@ -115,6 +119,8 @@ const customRender = ( ui, options = {} ) => {
 		registry,
 		history,
 		waitForRegistry: () => act( waitForRegistry ),
+		waitForResolvingPromise: ( delay ) =>
```
Similar to my other comment, the name is a bit ambiguous, and sounds more like our `AWAIT` action in the datastore which is essentially `async ( promise ) => promise`.

Ideally, we could enhance `waitForRegistry` to actually wait for everything as intended rather than introducing an arbitrary delay. Can we not simply advance the timers X amount before calling `await waitForRegistry()` to achieve the same?

techanvil replied:
As mentioned this has been removed/replaced :+1:

aaemnnosttv reviewed (changes requested):
Thanks @techanvil ! This is great to see finally approaching completion 😄 

This is looking good and overall it's great to see most of the changes are the same kind of thing throughout. At the same time, I wonder if there are changes we could make at a higher level to avoid some of that repetition?

As mentioned before, let's try to get this through sooner than later to take advantage of the longer dev cycle we have right now.

techanvil commented on assets/js/googlesitekit/datastore/user/surveys.test.js:
```diff
@@ -246,6 +258,8 @@ describe( 'core/user surveys', () => {
 					},
 				} );
+
+				jest.useRealTimers();
```
Good shout, I've removed this, it's no longer needed as I've moved the outer call to `jest.useRealTimers` to a `beforeEach` instead of a `beforeAll`.

techanvil commented on tests/js/jest.config.js:
```diff
@@ -36,6 +36,8 @@ module.exports = {
 	// Matches aliases in webpack.config.js.
 	moduleNameMapper: {
+		'^react$': '<rootDir>/node_modules/react',
+		'^react-dom$': '<rootDir>/node_modules/react-dom',
```
Thanks for highlighting this - I've been able to remove it, having now got rid of the multiple installed React instances that had crept in to `node_modules`.

techanvil commented on tests/js/setup-before-after.js:
```diff
@@ -32,6 +32,10 @@ import { enabledFeatures } from '../../assets/js/features';
 
+beforeAll( () => {
+	jest.useRealTimers();
+} );
```
I've added a comment about this. It would be more work and more problematic to rely on fake timers everywhere as we'd need to manually advance them for every resolver call.

https://github.com/google/site-kit-wp/blob/07e66cac79329d893ccbad2d11e2ecf21ecf9249/tests/js/setup-before-after.js#L38-L43

As for running pending timers, we already have that covered in a global `afterEach` -

https://github.com/google/site-kit-wp/blob/07e66cac79329d893ccbad2d11e2ecf21ecf9249/tests/js/setup-before-after.js#L64-L74

techanvil commented on webpack.config.js:
```diff
@@ -230,6 +230,7 @@ const createMinimizerRules = ( mode ) => [
 
 const resolve = {
 	alias: {
+		react: path.resolve( './node_modules/react' ),
```
Thanks for highlighting, this is gone too :)

techanvil commented on tests/js/test-utils.js:
```diff
@@ -115,6 +119,8 @@ const customRender = ( ui, options = {} ) => {
 		history,
 		waitForRegistry: () => act( waitForRegistry ),
+		waitForResolvingPromise: ( delay ) =>
+			act( () => resolvingPromise( delay ) ),
```
This has been replaced with `waitForRegistryAndTimeouts`, which is appropriate here as it uses `waitForRegistry`.

techanvil commented:
> Thanks @techanvil ! This is great to see finally approaching completion :smile:
> 
> This is looking good and overall it's great to see most of the changes are the same kind of thing throughout. At the same time, I wonder if there are changes we could make at a higher level to avoid some of that repetition?
> 
> As mentioned before, let's try to get this through sooner than later to take advantage of the longer dev cycle we have right now.

Thanks @aaemnnosttv, yep it's great to have this on the verge of completion! :crossed_fingers: 

Unfortunately it doesn't look like there's much room to avoid the repetition due to the nature of the change to `@wordpress/data`, we simply have to wait for resolvers in many new places. One thing we _could_ do is to provide a custom `it` function that adds an `await waitForTimeouts()` at the end of the test case, we could use that in place of the various tests which have been fixed by adding an await to the end of the test. I actually started down the path of [patching the global `it` function](https://github.com/google/site-kit-wp/pull/6203/commits/4d9ab837a7a623dd8e0d0d6e1e3627daeb39f12d) to do this, but reverted it as it was causing problems.

aaemnnosttv commented on tests/js/track-timeouts.js:
```diff
@@ -0,0 +1,74 @@
+ *
+ * @return {void}
+ */
+export function setupTimeoutTracker() {
+	let originalTimeoutAPI;
+
+	global.useTrackedTimeouts = () => {
+		originalTimeoutAPI = originalTimeoutAPI || {
+			setTimeout: global.setTimeout,
+			clearTimeout: global.clearTimeout,
+		};
```
I like the idea of this in principle but I'm not sure it's the right solution here. There are some concerns I have with this API and its coupling to whatever timer API is present at the time. Together, with the use of globals, this has the potential to be stacked on top of itself, as well as inconsistent results based on switching timer implementations with Jest.

Ideally, this would be something Jest would provide itself in order to provide a consistent experience with all timer implementations. From what I can tell, I don't think this is needed. See next comments.

techanvil replied:
I understand your reticence here, I think some of your concerns could be alleviated with a bit more work, but ultimately we can do without it and as you say it's something that would be nice to see Jest provide itself...

aaemnnosttv commented on tests/js/test-utils.js:
```diff
@@ -115,6 +115,11 @@ const customRender = ( ui, options = {} ) => {
 		registry,
 		history,
 		waitForRegistry: () => act( waitForRegistry ),
+		waitForRegistryAndTimeouts: () =>
+			act( async () => {
+				await waitForRegistry();
+				return global.waitForTimeouts();
+			} ),
```
Since the main problem here is (as you've identified) a [`setTimeout` in the registry itself](https://github.com/WordPress/gutenberg/blob/07baf5a12007d31bbd4ee22113b07952f7eacc26/packages/data/src/namespace-store/index.js#L296-L310), I think it makes sense to make the necessary change here to `waitForRegistry` rather than introducing a new utility. In my initial testing, it seems like all we need to do additionally await a `setTimeout` in that function, i.e.

```diff
diff --git a/tests/js/utils.js b/tests/js/utils.js
index f16ded614c..00e4762e19 100644
--- a/tests/js/utils.js
+++ b/tests/js/utils.js
@@ -503,7 +503,10 @@ export const createWaitForRegistry = ( registry ) => {
 	// We unsubscribe afterwards to allow for potential additions while
 	// Promise.all is resolving.
 	return async () => {
-		await Promise.all( updates );
+		await Promise.all( [
+			...updates,
+			new Promise( ( r ) => setTimeout( r ) ),
+		] );
 		unsubscribe();
 	};
 };
```

This resulted in 9 failed tests (which look potentially legit) when running all `datastore|components` tests and simply reverting the previous use of `waitForRegistry` instead of `waitForRegistryAndTimeouts` in those tests as well as disabling everything in `track-timeouts`.

The above diff could potentially be improved but I think it shows that we could be much closer to a simpler solution here than we thought.

techanvil replied:
I've updated `waitForRegistry` to wait for pending timeouts. I've added a 2ms delay here and elsewhere to avoid a race condition with the default timeouts - see the new [`waitForDefaultTimeouts`](https://github.com/google/site-kit-wp/blob/6f777e9daa57ccaf29455f23387ec76c6a04d290/tests/js/utils.js#L508) utility. Without this, some tests will still fail occasionally. I've documented this in the JSDoc for the `waitForDefaultTimeouts`.

I have been wondering if it should simply be called `waitForTimeouts`, due to the caveat mentioned in the JSDoc, what do you think?

https://github.com/google/site-kit-wp/blob/6f777e9daa57ccaf29455f23387ec76c6a04d290/tests/js/utils.js#L496-L498

aaemnnosttv reviewed (commented):
Thanks @techanvil – I think we're getting quite close here. I have a few points left to address regarding the final approach. Let me know if you have any questions!

techanvil commented on tests/e2e/config/bootstrap.js:
```diff
@@ -74,8 +74,8 @@ const pageEvents = [];
-// Set default timeout for individual expect-puppeteer assertions. (Default: 500)
-setDefaultOptions( { timeout: EXPECT_PUPPETEER_TIMEOUT || 500 } );
+// Set default timeout for individual expect-puppeteer assertions. (Default: 1000)
+setDefaultOptions( { timeout: EXPECT_PUPPETEER_TIMEOUT || 1000 } );
```
For some reason, the E2E WP 5.2 workflow started failing quite regularly on this branch, as can be seen here: https://github.com/google/site-kit-wp/actions/workflows/e2e-tests-wp-5-2.yml?query=branch%3Ainfra%2F1769-upgrade-wordpress-data

The failures are all in the same area, waiting for the page to navigate to the main Dashboard, e.g. https://github.com/google/site-kit-wp/actions/runs/3750192482/jobs/6370581519#step:12:267

![image](https://user-images.githubusercontent.com/18395600/208967763-9e9fdac3-81f6-4afe-b576-13eac78c8088.png)

I was able to reproduce this locally, and found increasing the default assertion timeout allowed the tests to run without fail.

I think we'd be well off increasing this globally here to help make our tests a bit more robust, rather than on a per-test or per-assertion basis.

techanvil commented:
> Thanks @techanvil – I think we're getting quite close here. I have a few points left to address regarding the final approach. Let me know if you have any questions!

Thanks @aaemnnosttv. I've addressed your points, and also addressed some E2E test flakiness that had emerged. Hopefully we are on the home stretch now :)

aaemnnosttv reviewed (approved):
LGTM, excellent work @techanvil ! 🎉 🚀
